### PR TITLE
refactor(rect-grid): refactor table column reordering

### LINKED ISSUE #154: Table column reordering is not working with Virtual Table
Currently, it's impossible to drop a dragged column on a Virtual Table. It does nothing.

### PR DESCRIPTION
**Pull request by gsobolev**
fixes #154 

BREAKING CHANGES:

- The `ColumnOrderState` plugin has been renamed to `TableColumnReordering` and is now available via the `@devexpress/dx-react-grid-bootstrap3` and `@devexpress/dx-react-grid-material-ui` packages.
- The `TableView` plugin's `allowColumnReordering` property has been removed and the `TableColumnReordering` plugin now depends on the `TableView` plugin. Thus, it is enough to link the `TableColumnReordering` plugin below the `TableView` plugin to enable column reordering.

Before:
```jsx
import {
  // ...
  ColumnOrderState
} from '@devexpress/dx-react-grid';

// ...

<ColumnOrderState defaultOrder={[/* ... */]} />
<TableView allowColumnReordering />
```

After:
```jsx
import {
  // ...
  TableColumnReordering
} from '@devexpress/dx-react-grid-bootstrap3';
// } from '@devexpress/dx-react-grid-material-ui';

// ...

<TableView />
<TableColumnReordering defaultOrder={[/* ... */]} />
```

### CODE REVIEW

File: packages/dx-react-grid/docs/guides/column-reordering.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ The Grid component allows reordering grid columns programmatically or using the
 
 The following plugins implement column reordering features:
 
-- [ColumnOrderState](../reference/column-order-state.md) - controls the column order
+- [TableColumnReordering](../reference/table-column-reordering.md) - controls the column order
```
Comment (kvet) [inline packages/dx-react-grid/docs/guides/column-reordering.md]: Place it after TableView

File: packages/dx-react-grid/docs/guides/column-reordering.md [hunk 2]
```diff
@@ -15,21 +15,18 @@ Note that the [plugin order](./plugin-overview.md#plugin-order) is important.
 
 ## Column Reordering Setup
 
-Import the plugins listed above to set up a simple Grid with column reordering enabled.
+Import the plugins listed above to set up a simple Grid with column reordering enabled. Note that the [TableColumnReordering](../reference/table-column-reordering.md) plugin should be placed after the [TableView](../reference/table-view.md) plugin.
```
Comment (kvet) [inline packages/dx-react-grid/docs/guides/column-reordering.md]: Not sure that this note should be there. Seems like it should be listed in breaking change description.

File: packages/dx-react-grid/docs/reference/table-view.md [hunk 3]
```diff
@@ -6,7 +6,7 @@ This plugin renders Grid data as a table. It contains the Table View and Table V
 
 ### Dependencies
 
-- [ColumnOrderState](column-order-state.md) [Optional]
+- [TableColumnReordering](table-column-reordering.md) [Optional]
```
Comment (kvet) [inline packages/dx-react-grid/docs/reference/table-view.md]: Not sure that this is true...

File: packages/dx-react-grid/docs/reference/table-view.md [hunk 4]
```diff
@@ -21,7 +21,6 @@ tableNoDataCellTemplate | (args: [TableCellArgs](#table-cell-args)) => ReactElem
 tableNoDataRowTemplate | (args: [TableRowArgs](#table-row-args)) => ReactElement | | Renders a table row using the specified parameters when the table is empty.
 tableStubCellTemplate | (args: [TableCellArgs](#table-cell-args)) => ReactElement | | Renders a stub table cell if the cell value is not provided.
 tableStubHeaderCellTemplate | (args: [TableCellArgs](#table-cell-args)) => ReactElement | | Renders a stub header cell if the cell value is not provided.
-allowColumnReordering | boolean | false | If true, it allows end-users to change the column's order by dragging it. Requires the [ColumnOrderState](column-order-state.md) and the [DragDropContext](drag-drop-context.md) dependencies.
```
Comment (kvet) [inline packages/dx-react-grid/docs/reference/table-view.md]: Is it listed in BC description?

File: packages/dx-react-grid/src/plugins/table-column-reordering.jsx [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+    return (
+      <PluginContainer
+        pluginName="TableColumnReordering"
+        dependencies={pluginDependencies}
+      >
+        <Getter name="tableColumns" computed={columnsComputed} />
+        <Getter name="tableHeaderRows" computed={tableHeaderRowsComputed} />
+        <Template name="tableView">
+          <DropTarget
```
Comment (kvet) [inline packages/dx-react-grid/src/plugins/table-column-reordering.jsx]: It is part of the web components. It should not be present in core plugin. Please, extract an additional template

File: packages/dx-react-grid/docs/guides/column-reordering.md (hunk 2 above)
Comment (gsobolev) [reply packages/dx-react-grid/docs/guides/column-reordering.md]: I'm going to edit the list of breaking changes anyway. But this is the guide, so the note may be useful for the first coming users. That is why I added it.

File: packages/dx-react-grid-bootstrap3/src/templates/virtual-table/virtual-box.jsx [hunk 6]
```diff
@@ -29,6 +29,7 @@ function getVisibleItems(options) {
     const itemInfo = options.itemInfo(index);
     const itemSize = itemInfo.size;
     const itemStick = itemInfo.stick;
+    const key = itemInfo.key;
```
Comment (SergeyAlexeev) [inline packages/dx-react-grid-bootstrap3/src/templates/virtual-table/virtual-box.jsx]: Does eslint allow to use this construction? I mean, should we use object decomposition instead?

File: packages/dx-react-grid/docs/reference/table-column-reordering.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+### <a name="reordering-cell-args"></a>ReorderingCellArgs
+
+Describes properties passed to the reordering row cell template.
+
+A value with the [TableCellArgs](table-view.md#table-cell-args) shape extended by the following fields:
+
+Field | Type | Description
+------|------|------------
+getCellDimension | (fn: () => { left: number, right: number }) => void) | A function that accepts another function responsible for getting the cell's horizontal bounds.
```
Comment (SergeyAlexeev) [inline packages/dx-react-grid/docs/reference/table-column-reordering.md]: Is `fn` a final name?)

File: packages/dx-react-grid-bootstrap3/src/templates/virtual-table/virtual-box.jsx (hunk 6 above)
Comment (gsobolev) [reply packages/dx-react-grid-bootstrap3/src/templates/virtual-table/virtual-box.jsx]: There were no errors. But it looks like _eslint_ will fail after the #431 merge.

File: packages/dx-react-grid/docs/reference/table-column-reordering.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+- [TableView](table-view.md)
+
+### Properties
+
+Name | Type | Default | Description
+-----|------|---------|------------
+order | Array&lt;string&gt; | | Specifies the column order.
+defaultOrder | Array&lt;string&gt; | | Specifies the initial column order in the uncontrolled mode.
+onOrderChange | (nextOrder: Array&lt;string&gt;) => void | | Handles column order changes.
+tableViewTemplate | (args: [TableViewArgs](#table-view-args)) => ReactElement | | A component that renders a wrapper, which is responsible for column dropping, around the table.
```
Comment (kvet) [inline packages/dx-react-grid/docs/reference/table-column-reordering.md]: Maybe call it `tableContainerTemplate` or `tableReorderingContainerTemplate`?

File: packages/dx-react-grid/docs/reference/table-column-reordering.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+### <a name="table-view-args"></a>TableViewArgs
+
+Describes properties passed to the table view template.
+
+A value with the following shape:
+
+Field | Type | Description
+------|------|------------
+onOver | (args: { payload: Array&lt;{ columnName: string }&gt;], clientOffset: { x: number } }) => void | Handles column drag over.
```
Comment (kvet) [inline packages/dx-react-grid/docs/reference/table-column-reordering.md]: There is a redundant symbol: `]`.

File: packages/dx-react-grid/docs/guides/column-reordering.md (hunk 2 above)
Comment (kvet) [reply packages/dx-react-grid/docs/guides/column-reordering.md]: Typically, we add a note about plugin order importance. And assume that this info is in reference. Let's write consistent guides.

File: packages/dx-react-grid-bootstrap3/src/templates/virtual-table.jsx [hunk 10]
```diff
@@ -33,8 +33,8 @@ export class VirtualTable extends React.Component {
     };
 
     this.rowHeight = (row) => {
-      let height = row.height || DEFAULT_HEIGHT;
-      if (height === 'auto') {
+      let { height } = row;
+      if (typeof (row.height) === 'undefined' || height === 'auto') {
```
Comment (SergeyAlexeev) [inline packages/dx-react-grid-bootstrap3/src/templates/virtual-table.jsx]: `typeof (row.height) === 'undefined'` => `typeof height === 'undefined'`

File: packages/dx-react-grid/src/components/table-layout.jsx [hunk 11]
```diff
@@ -3,150 +3,68 @@
...
-
 import { RowsBlockLayout } from './table-layout/rows-block-layout';
 
 const TABLE_FLEX_TYPE = 'flex';
 
+const areColumnsChanged = (prevColumns, nextColumns) => {
+  if (prevColumns.length !== nextColumns.length) return true;
+  const prevKeys = prevColumns.map(column => column.key);
+  return nextColumns.filter(column =>
+    prevKeys.indexOf(column.key) === -1).length !== 0;
```
Comment (SergeyAlexeev) [inline packages/dx-react-grid/src/components/table-layout.jsx]: Can we use `Array,find` or `Array.findIndex` here? It doesn't iterate all array items.
```js
return nextColumns.find(column => prevKeys.indexOf(column.key) === -1) !== undefined;
```

File: packages/dx-react-grid/src/components/table-layout.jsx [hunk 12]
```diff
@@ -3,150 +3,68 @@
...
 } from '@devexpress/dx-grid-core';
-
 import { RowsBlockLayout } from './table-layout/rows-block-layout';
 
 const TABLE_FLEX_TYPE = 'flex';
 
+const areColumnsChanged = (prevColumns, nextColumns) => {
+  if (prevColumns.length !== nextColumns.length) return true;
+  const prevKeys = prevColumns.map(column => column.key);
+  return nextColumns.filter(column =>
```
Comment (SergeyAlexeev) [inline packages/dx-react-grid/src/components/table-layout.jsx]: Can we use `Array.find` or `Array.findIndex` here? It doesn't iterate all array items.
```js
return nextColumns.find(column => prevKeys.indexOf(column.key) === -1) !== undefined;
```

File: packages/dx-react-grid/src/components/table-layout/row-layout.test.jsx [hunk 13]
```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import { TABLE_DATA_TYPE, getTableRowColumnsWithColSpan } from '@devexpress/dx-grid-core';
+import { setupConsole } from '@devexpress/dx-testing';
+import { RowLayout } from './row-layout';
+
+jest.mock('@devexpress/dx-grid-core', () => ({
+  TABLE_DATA_TYPE: 'd',
```
Comment (kvet) [inline packages/dx-react-grid/src/components/table-layout/row-layout.test.jsx]: De we really need to use the `TABLE_DATA_TYPE` constant in tests?

**Pull request merged**